{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e32b8cc4",
   "metadata": {},
   "source": [
    "### Demonstration Overview: ClinVar to FHIR Allele Profiles\n",
    "\n",
    "This notebook demonstrates how to extract, explore, and translate ClinVar variation data into FHIR Allele Profiles, optimized for execution in a GitHub Codespace or local environment.\n",
    "    \n",
    "#### Key features of this notebook:\n",
    "- **Stream large datasets efficiently:** Read and process a compressed .jsonl.gz ClinVar file line-by-line without loading it fully into memory.\n",
    "- **Filter relevant records:** Identify and retain entries where `members` include an `Allele` type.\n",
    "- **Preview representative examples:** Display either the full ClinVar record, the embedded VRS Allele object, or both, depending on the selected display mode.\n",
    "- **Extract VRS Alleles:** Collect a small, reproducible subset (e.g., 5 examples) of VRS Allele objects for quick translation.\n",
    "- **Perform translation:** Convert those VRS Allele objects into FHIR Allele Profiles to demonstrate the end-to-end mapping workflow.\n",
    "- **Portable execution:** Runs efficiently in a GitHub Codespace or on a standard laptop without requiring additional compute resources.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38043ff9",
   "metadata": {},
   "source": [
    "> **Note:**  \n",
    "> - We have translated a **subset** of the original ClinVar data folder, generating **FHIR Allele Profiles for approximately 20,000 examples**, and are now working toward **translating the complete dataset**.\n",
    "> - If you would like access to the **input or output data files** used in this demonstration, please contact the code owners.  \n",
    "> - A **command-line Python script** is currently under development to perform this same ClinVar -> FHIR Allele Profile translation. Once released, it will be available in the `src/utils/` directory of this repository."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94fb6330",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip, json, random\n",
    "from ga4gh.vrs.models import Allele\n",
    "from translators.vrs_to_fhir import VrsToFhirAlleleTranslator \n",
    "from ga4gh.vrs.dataproxy import create_dataproxy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "355e079a",
   "metadata": {},
   "outputs": [],
   "source": [
    "dp = create_dataproxy(uri=\"seqrepo+file:///usr/local/share/seqrepo/2024-12-20\")\n",
    "\n",
    "vrs_translator = VrsToFhirAlleleTranslator(dp=dp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "dbe517f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "path = \"data/clinvar_gks_variation_2025_09_28_v2_4_3.jsonl.gz\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08c03b7f",
   "metadata": {},
   "source": [
    "### Filtering for Allele Records\n",
    "We filter for records where \"members\" include an object with \"type\": \"Allele\".\n",
    "Only these entries contain GA4GH VRS Allele representations, which are required for downstream translation into FHIR Allele Profiles.\n",
    "\n",
    "This filtering step is performed immediately after obtaining the ClinVar data to ensure that we only process records that can be represented as VRS Alleles."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "25258473",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----------- Original Example -----------\n",
      "{\n",
      "  \"constraints\": [\n",
      "    {\n",
      "      \"allele\": \"2/members/0/\",\n",
      "      \"relations\": [\n",
      "        {\n",
      "          \"primaryCoding\": {\n",
      "            \"code\": \"liftover_to\",\n",
      "            \"system\": \"ga4gh-gks-term:allele-relation\"\n",
      "          }\n",
      "        },\n",
      "        {\n",
      "          \"primaryCoding\": {\n",
      "            \"code\": \"transcribed_to\",\n",
      "            \"iris\": [\n",
      "              \"http://www.sequenceontology.org/browser/current_release/term/transcribed_to\"\n",
      "            ],\n",
      "            \"system\": \"http://www.sequenceontology.org\"\n",
      "          }\n",
      "        }\n",
      "      ],\n",
      "      \"type\": \"DefiningAlleleConstraint\"\n",
      "    }\n",
      "  ],\n",
      "  \"extensions\": [\n",
      "    {\n",
      "      \"name\": \"clinvarHgvsList\",\n",
      "      \"value\": [\n",
      "        {\n",
      "          \"maneSelect\": true,\n",
      "          \"nucleotideExpression\": {\n",
      "            \"syntax\": \"hgvs.n\",\n",
      "            \"value\": \"NR_023343.3:n.16G>C\"\n",
      "          },\n",
      "          \"nucleotideType\": \"non-coding\"\n",
      "        },\n",
      "        {\n",
      "          \"molecularConsequence\": [\n",
      "            {\n",
      "              \"code\": \"SO:0001627\",\n",
      "              \"iris\": [\n",
      "                \"https://identifiers.org/SO:0001627\"\n",
      "              ],\n",
      "              \"name\": \"intron_variant\",\n",
      "              \"system\": \"http://www.sequenceontology.org/\"\n",
      "            }\n",
      "          ],\n",
      "          \"nucleotideExpression\": {\n",
      "            \"syntax\": \"hgvs.c\",\n",
      "            \"value\": \"NM_001378004.1:c.196-570C>G\"\n",
      "          },\n",
      "          \"nucleotideType\": \"coding\"\n",
      "        },\n",
      "        {\n",
      "          \"molecularConsequence\": [\n",
      "            {\n",
      "              \"code\": \"SO:0001627\",\n",
      "              \"iris\": [\n",
      "                \"https://identifiers.org/SO:0001627\"\n",
      "              ],\n",
      "              \"name\": \"intron_variant\",\n",
      "              \"system\": \"http://www.sequenceontology.org/\"\n",
      "            }\n",
      "          ],\n",
      "          \"nucleotideExpression\": {\n",
      "            \"syntax\": \"hgvs.c\",\n",
      "            \"value\": \"NM_001207051.2:c.196-570C>G\"\n",
      "          },\n",
      "          \"nucleotideType\": \"coding\"\n",
      "        },\n",
      "        {\n",
      "          \"nucleotideExpression\": {\n",
      "            \"syntax\": \"hgvs.g\",\n",
      "            \"value\": \"NC_000002.12:g.121530895G>C\"\n",
      "          },\n",
      "          \"nucleotideType\": \"genomic, top-level\"\n",
      "        },\n",
      "        {\n",
      "          \"molecularConsequence\": [\n",
      "            {\n",
      "              \"code\": \"SO:0001627\",\n",
      "              \"iris\": [\n",
      "                \"https://identifiers.org/SO:0001627\"\n",
      "              ],\n",
      "              \"name\": \"intron_variant\",\n",
      "              \"system\": \"http://www.sequenceontology.org/\"\n",
      "            }\n",
      "          ],\n",
      "          \"nucleotideExpression\": {\n",
      "            \"syntax\": \"hgvs.c\",\n",
      "            \"value\": \"NM_015282.3:c.196-570C>G\"\n",
      "          },\n",
      "          \"nucleotideType\": \"coding\"\n",
      "        },\n",
      "        {\n",
      "          \"maneSelect\": true,\n",
      "          \"molecularConsequence\": [\n",
      "            {\n",
      "              \"code\": \"SO:0001627\",\n",
      "              \"iris\": [\n",
      "                \"https://identifiers.org/SO:0001627\"\n",
      "              ],\n",
      "              \"name\": \"intron_variant\",\n",
      "              \"system\": \"http://www.sequenceontology.org/\"\n",
      "            }\n",
      "          ],\n",
      "          \"nucleotideExpression\": {\n",
      "            \"syntax\": \"hgvs.c\",\n",
      "            \"value\": \"NM_001395891.1:c.196-570C>G\"\n",
      "          },\n",
      "          \"nucleotideType\": \"coding\"\n",
      "        },\n",
      "        {\n",
      "          \"molecularConsequence\": [\n",
      "            {\n",
      "              \"code\": \"SO:0001627\",\n",
      "              \"iris\": [\n",
      "                \"https://identifiers.org/SO:0001627\"\n",
      "              ],\n",
      "              \"name\": \"intron_variant\",\n",
      "              \"system\": \"http://www.sequenceontology.org/\"\n",
      "            }\n",
      "          ],\n",
      "          \"nucleotideExpression\": {\n",
      "            \"syntax\": \"hgvs.c\",\n",
      "            \"value\": \"NM_001378005.1:c.196-570C>G\"\n",
      "          },\n",
      "          \"nucleotideType\": \"coding\"\n",
      "        },\n",
      "        {\n",
      "          \"nucleotideExpression\": {\n",
      "            \"syntax\": \"hgvs.g\",\n",
      "            \"value\": \"NC_000002.11:g.122288471G>C\"\n",
      "          },\n",
      "          \"nucleotideType\": \"genomic, top-level\"\n",
      "        },\n",
      "        {\n",
      "          \"nucleotideExpression\": {\n",
      "            \"syntax\": \"hgvs.g\",\n",
      "            \"value\": \"LRG_1202:g.5016G>C\"\n",
      "          },\n",
      "          \"nucleotideType\": \"genomic\"\n",
      "        },\n",
      "        {\n",
      "          \"molecularConsequence\": [\n",
      "            {\n",
      "              \"code\": \"SO:0001627\",\n",
      "              \"iris\": [\n",
      "                \"https://identifiers.org/SO:0001627\"\n",
      "              ],\n",
      "              \"name\": \"intron_variant\",\n",
      "              \"system\": \"http://www.sequenceontology.org/\"\n",
      "            }\n",
      "          ],\n",
      "          \"nucleotideExpression\": {\n",
      "            \"syntax\": \"hgvs.c\",\n",
      "            \"value\": \"NM_001378003.1:c.196-570C>G\"\n",
      "          },\n",
      "          \"nucleotideType\": \"coding\"\n",
      "        },\n",
      "        {\n",
      "          \"nucleotideExpression\": {\n",
      "            \"syntax\": \"hgvs.n\",\n",
      "            \"value\": \"LRG_1202t1:n.16G>C\"\n",
      "          },\n",
      "          \"nucleotideType\": \"coding\"\n",
      "        },\n",
      "        {\n",
      "          \"nucleotideExpression\": {\n",
      "            \"syntax\": \"hgvs.g\",\n",
      "            \"value\": \"NG_029832.1:g.5016G>C\"\n",
      "          },\n",
      "          \"nucleotideType\": \"genomic\"\n",
      "        },\n",
      "        {\n",
      "          \"molecularConsequence\": [\n",
      "            {\n",
      "              \"code\": \"SO:0001627\",\n",
      "              \"iris\": [\n",
      "                \"https://identifiers.org/SO:0001627\"\n",
      "              ],\n",
      "              \"name\": \"intron_variant\",\n",
      "              \"system\": \"http://www.sequenceontology.org/\"\n",
      "            }\n",
      "          ],\n",
      "          \"nucleotideExpression\": {\n",
      "            \"syntax\": \"hgvs.c\",\n",
      "            \"value\": \"NM_001142273.2:c.196-570C>G\"\n",
      "          },\n",
      "          \"nucleotideType\": \"coding\"\n",
      "        },\n",
      "        {\n",
      "          \"molecularConsequence\": [\n",
      "            {\n",
      "              \"code\": \"SO:0001627\",\n",
      "              \"iris\": [\n",
      "                \"https://identifiers.org/SO:0001627\"\n",
      "              ],\n",
      "              \"name\": \"intron_variant\",\n",
      "              \"system\": \"http://www.sequenceontology.org/\"\n",
      "            }\n",
      "          ],\n",
      "          \"nucleotideExpression\": {\n",
      "            \"syntax\": \"hgvs.c\",\n",
      "            \"value\": \"NM_001142274.2:c.196-570C>G\"\n",
      "          },\n",
      "          \"nucleotideType\": \"coding\"\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"name\": \"categoricalVariationType\",\n",
      "      \"value\": \"CanonicalAllele\"\n",
      "    },\n",
      "    {\n",
      "      \"name\": \"clinvarGeneList\",\n",
      "      \"value\": [\n",
      "        {\n",
      "          \"entrez_gene_id\": \"100151683\",\n",
      "          \"hgnc_id\": \"HGNC:34016\",\n",
      "          \"iris\": [\n",
      "            \"https://identifiers.org/hgnc:34016\",\n",
      "            \"https://identifiers.org/ncbigene:100151683\",\n",
      "            \"https://www.ncbi.nlm.nih.gov/gene/100151683\"\n",
      "          ],\n",
      "          \"relationship_type\": \"within multiple genes by overlap\",\n",
      "          \"source\": \"submitted\",\n",
      "          \"symbol\": \"RNU4ATAC\"\n",
      "        },\n",
      "        {\n",
      "          \"entrez_gene_id\": \"23332\",\n",
      "          \"hgnc_id\": \"HGNC:17088\",\n",
      "          \"iris\": [\n",
      "            \"https://identifiers.org/hgnc:17088\",\n",
      "            \"https://identifiers.org/ncbigene:23332\",\n",
      "            \"https://www.ncbi.nlm.nih.gov/gene/23332\"\n",
      "          ],\n",
      "          \"relationship_type\": \"within multiple genes by overlap\",\n",
      "          \"source\": \"calculated\",\n",
      "          \"symbol\": \"CLASP1\"\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"name\": \"clinvarVariationType\",\n",
      "      \"value\": \"single nucleotide variant\"\n",
      "    },\n",
      "    {\n",
      "      \"name\": \"definingVrsVariationType\",\n",
      "      \"value\": \"Allele\"\n",
      "    },\n",
      "    {\n",
      "      \"name\": \"clinvarCytogeneticLocation\",\n",
      "      \"value\": \"2q14.2\"\n",
      "    },\n",
      "    {\n",
      "      \"name\": \"clinvarSubclassType\",\n",
      "      \"value\": \"SimpleAllele\"\n",
      "    }\n",
      "  ],\n",
      "  \"id\": \"clinvar:1469262\",\n",
      "  \"mappings\": [\n",
      "    {\n",
      "      \"coding\": {\n",
      "        \"code\": \"1469262\",\n",
      "        \"iris\": [\n",
      "          \"https://identifiers.org/clinvar:1469262\"\n",
      "        ],\n",
      "        \"system\": \"ClinVar\"\n",
      "      },\n",
      "      \"relation\": \"exactMatch\"\n",
      "    },\n",
      "    {\n",
      "      \"coding\": {\n",
      "        \"code\": \"CA428887824\",\n",
      "        \"iris\": [\n",
      "          \"https://reg.clinicalgenome.org/redmine/projects/registry/genboree_registry/by_caid?caid=CA428887824\"\n",
      "        ],\n",
      "        \"system\": \"ClinGen\"\n",
      "      },\n",
      "      \"relation\": \"closeMatch\"\n",
      "    },\n",
      "    {\n",
      "      \"coding\": {\n",
      "        \"code\": \"750325275\",\n",
      "        \"iris\": [\n",
      "          \"https://identifiers.org/dbsnp:rs750325275\"\n",
      "        ],\n",
      "        \"system\": \"dbSNP\"\n",
      "      },\n",
      "      \"relation\": \"relatedMatch\"\n",
      "    }\n",
      "  ],\n",
      "  \"members\": [\n",
      "    {\n",
      "      \"digest\": \"RYY2yzWCjihuu2hmqFWu8qg7aUB3aPwR\",\n",
      "      \"expressions\": [\n",
      "        {\n",
      "          \"syntax\": \"spdi\",\n",
      "          \"value\": \"NC_000002.12:121530894:G:C\"\n",
      "        },\n",
      "        {\n",
      "          \"syntax\": \"hgvs.g\",\n",
      "          \"value\": \"NC_000002.12:g.121530895G>C\"\n",
      "        },\n",
      "        {\n",
      "          \"syntax\": \"gnomad\",\n",
      "          \"value\": \"2-121530895-G-C\"\n",
      "        }\n",
      "      ],\n",
      "      \"id\": \"ga4gh:VA.RYY2yzWCjihuu2hmqFWu8qg7aUB3aPwR\",\n",
      "      \"location\": {\n",
      "        \"digest\": \"3sVPpjle__QNFe8yGfLEja0R42VIZIp8\",\n",
      "        \"end\": 121530895,\n",
      "        \"id\": \"ga4gh:SL.3sVPpjle__QNFe8yGfLEja0R42VIZIp8\",\n",
      "        \"sequenceReference\": {\n",
      "          \"extensions\": [\n",
      "            {\n",
      "              \"name\": \"assembly\",\n",
      "              \"value\": \"GRCh38\"\n",
      "            }\n",
      "          ],\n",
      "          \"moleculeType\": \"genomic\",\n",
      "          \"name\": \"NC_000002.12\",\n",
      "          \"refgetAccession\": \"SQ.pnAqCRBrTsUoBghSD1yp_jXWSmlbdh4g\",\n",
      "          \"residueAlphabet\": \"na\",\n",
      "          \"type\": \"SequenceReference\"\n",
      "        },\n",
      "        \"start\": 121530894,\n",
      "        \"type\": \"SequenceLocation\"\n",
      "      },\n",
      "      \"name\": \"NC_000002.12:g.121530895G>C\",\n",
      "      \"state\": {\n",
      "        \"sequence\": \"C\",\n",
      "        \"type\": \"LiteralSequenceExpression\"\n",
      "      },\n",
      "      \"type\": \"Allele\"\n",
      "    }\n",
      "  ],\n",
      "  \"name\": \"NC_000002.12:g.121530895G>C\",\n",
      "  \"type\": \"CategoricalVariant\"\n",
      "}\n",
      "----------- VRS Example from Original -----------\n",
      "{\n",
      "  \"digest\": \"RYY2yzWCjihuu2hmqFWu8qg7aUB3aPwR\",\n",
      "  \"expressions\": [\n",
      "    {\n",
      "      \"syntax\": \"spdi\",\n",
      "      \"value\": \"NC_000002.12:121530894:G:C\"\n",
      "    },\n",
      "    {\n",
      "      \"syntax\": \"hgvs.g\",\n",
      "      \"value\": \"NC_000002.12:g.121530895G>C\"\n",
      "    },\n",
      "    {\n",
      "      \"syntax\": \"gnomad\",\n",
      "      \"value\": \"2-121530895-G-C\"\n",
      "    }\n",
      "  ],\n",
      "  \"id\": \"ga4gh:VA.RYY2yzWCjihuu2hmqFWu8qg7aUB3aPwR\",\n",
      "  \"location\": {\n",
      "    \"digest\": \"3sVPpjle__QNFe8yGfLEja0R42VIZIp8\",\n",
      "    \"end\": 121530895,\n",
      "    \"id\": \"ga4gh:SL.3sVPpjle__QNFe8yGfLEja0R42VIZIp8\",\n",
      "    \"sequenceReference\": {\n",
      "      \"extensions\": [\n",
      "        {\n",
      "          \"name\": \"assembly\",\n",
      "          \"value\": \"GRCh38\"\n",
      "        }\n",
      "      ],\n",
      "      \"moleculeType\": \"genomic\",\n",
      "      \"name\": \"NC_000002.12\",\n",
      "      \"refgetAccession\": \"SQ.pnAqCRBrTsUoBghSD1yp_jXWSmlbdh4g\",\n",
      "      \"residueAlphabet\": \"na\",\n",
      "      \"type\": \"SequenceReference\"\n",
      "    },\n",
      "    \"start\": 121530894,\n",
      "    \"type\": \"SequenceLocation\"\n",
      "  },\n",
      "  \"name\": \"NC_000002.12:g.121530895G>C\",\n",
      "  \"state\": {\n",
      "    \"sequence\": \"C\",\n",
      "    \"type\": \"LiteralSequenceExpression\"\n",
      "  },\n",
      "  \"type\": \"Allele\"\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "# Choose which variation you want to view.\n",
    "# Options:\n",
    "# \"full\" -> show the entire original JSON object\n",
    "# \"vrs\" -> show only the first VRS Allele example\n",
    "# \"both\" -> show both the full object and its VRS example\n",
    "# Optoions: \"Original Example: full, VRS Examples: vrs, or both\" \n",
    "\n",
    "display_mode = \"both\" \n",
    "\n",
    "with gzip.open(path, \"rt\", encoding=\"utf-8\") as f:\n",
    "    for line in f:\n",
    "        obj = json.loads(line)\n",
    "        members = obj.get(\"members\",[])\n",
    "        has_allele = False\n",
    "\n",
    "        for member in members: \n",
    "            if isinstance(member,dict) and member.get(\"type\")==\"Allele\":\n",
    "                has_allele = True\n",
    "                break\n",
    "\n",
    "        if has_allele:\n",
    "            if display_mode == \"full\":\n",
    "                 print(json.dumps(obj, indent=2))\n",
    "            elif display_mode == \"vrs\":\n",
    "\n",
    "                print(json.dumps(obj[\"members\"][0], indent=2))\n",
    "\n",
    "            elif display_mode == \"both\":\n",
    "                print(\"----------- Original Example -----------\")\n",
    "                print(json.dumps(obj, indent=2))\n",
    "                print(\"----------- VRS Example from Original -----------\")\n",
    "                print(json.dumps(obj[\"members\"][0], indent=2))\n",
    "            break\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46b6edb8",
   "metadata": {},
   "source": [
    "### Extracting Sample VRS Allele Examples\n",
    "- In this step, we extract five VRS Allele examples from the filtered ClinVar data to explore and test the translation process.\n",
    "- We use random.seed() to ensure that the same subset of examples is selected each time the notebook is run."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4f2104a",
   "metadata": {},
   "source": [
    "> **Note:**  \n",
    "> - Some entries in the dataset contain `\"type\": \"Allele\"` but cannot be successfully represented as `Allele(**)` objects.  \n",
    "> - These cases typically fail VRS validation because their `\"molecularType\"` is incorrectly set to `\"mrna\"` instead of `\"mRNA\"`.  \n",
    "> - This inconsistency may have been introduced during the **data input process** or during **translation into the dataset**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a510687a",
   "metadata": {},
   "outputs": [],
   "source": [
    "random.seed(42) \n",
    "vrs_examples = []\n",
    "\n",
    "with gzip.open(path, \"rt\", encoding=\"utf-8\") as f:\n",
    "    for line in f:\n",
    "        obj = json.loads(line)\n",
    "        members = obj.get('members',[])\n",
    "\n",
    "        for member in members:\n",
    "            if isinstance(member,dict) and member.get(\"type\")==\"Allele\":\n",
    "                vrs_examples.append(Allele(**member))\n",
    "                break\n",
    "        if len(vrs_examples) == 5:\n",
    "            break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "912a5d14",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'id': 'ga4gh:VA.RYY2yzWCjihuu2hmqFWu8qg7aUB3aPwR',\n",
       " 'type': 'Allele',\n",
       " 'name': 'NC_000002.12:g.121530895G>C',\n",
       " 'digest': 'RYY2yzWCjihuu2hmqFWu8qg7aUB3aPwR',\n",
       " 'expressions': [{'syntax': 'spdi', 'value': 'NC_000002.12:121530894:G:C'},\n",
       "  {'syntax': 'hgvs.g', 'value': 'NC_000002.12:g.121530895G>C'},\n",
       "  {'syntax': 'gnomad', 'value': '2-121530895-G-C'}],\n",
       " 'location': {'id': 'ga4gh:SL.3sVPpjle__QNFe8yGfLEja0R42VIZIp8',\n",
       "  'type': 'SequenceLocation',\n",
       "  'digest': '3sVPpjle__QNFe8yGfLEja0R42VIZIp8',\n",
       "  'sequenceReference': {'type': 'SequenceReference',\n",
       "   'name': 'NC_000002.12',\n",
       "   'extensions': [{'name': 'assembly', 'value': 'GRCh38'}],\n",
       "   'refgetAccession': 'SQ.pnAqCRBrTsUoBghSD1yp_jXWSmlbdh4g',\n",
       "   'residueAlphabet': 'na',\n",
       "   'moleculeType': 'genomic'},\n",
       "  'start': 121530894,\n",
       "  'end': 121530895},\n",
       " 'state': {'type': 'LiteralSequenceExpression', 'sequence': 'C'}}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Display one example VRS Allele to show its structure.\n",
    "vrs_examples[0].model_dump(exclude_none = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a610ed98",
   "metadata": {},
   "source": [
    "### Translating VRS Alleles into FHIR Allele Profiles\n",
    "\n",
    "- Translate the five extracted VRS Allele examples into FHIR Allele Profiles.\n",
    "- This step demonstrates the end-to-end VRS -> FHIR mapping process, on a small subset for quick, reproducible testing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1e24e698",
   "metadata": {},
   "outputs": [],
   "source": [
    "fhir_examples = []\n",
    "for vrs in vrs_examples:\n",
    "    fhir_examples.append(vrs_translator.translate_allele_to_fhir(vrs))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e09c4087",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'resourceType': 'MolecularDefinition',\n",
       " 'contained': [{'resourceType': 'MolecularDefinition',\n",
       "   'id': 'vrs-location-sequenceReference',\n",
       "   'extension': [{'url': 'https://w3id.org/ga4gh/schema/vrs/2.0.1/json/SequenceReference#properties/name',\n",
       "     'valueString': 'NC_000002.12'},\n",
       "    {'extension': [{'url': 'https://github.com/ga4gh/gks-core/blob/1.0/schema/gks-core/json/Extension#properties/name',\n",
       "       'valueString': 'assembly'},\n",
       "      {'url': 'https://github.com/ga4gh/gks-core/blob/1.0/schema/gks-core/json/Extension#properties/value',\n",
       "       'valueString': 'GRCh38'}]}],\n",
       "   'moleculeType': {'coding': [{'system': 'https://w3id.org/ga4gh/schema/vrs/2.0.1/json/SequenceReference#properties/moleculeType',\n",
       "      'code': 'genomic'}]},\n",
       "   'representation': [{'code': [{'coding': [{'system': 'https://w3id.org/ga4gh/schema/vrs/2.0.1/json/SequenceReference#properties/refgetAccession',\n",
       "         'code': 'SQ.pnAqCRBrTsUoBghSD1yp_jXWSmlbdh4g'}]}]}]}],\n",
       " 'identifier': [{'system': 'https://w3id.org/ga4gh/schema/vrs/2.0.1/json/Allele#properties/id',\n",
       "   'value': 'ga4gh:VA.RYY2yzWCjihuu2hmqFWu8qg7aUB3aPwR'},\n",
       "  {'system': 'https://w3id.org/ga4gh/schema/vrs/2.0.1/json/Allele#properties/name',\n",
       "   'value': 'NC_000002.12:g.121530895G>C'},\n",
       "  {'system': 'https://w3id.org/ga4gh/schema/vrs/2.0.1/json/Allele#properties/digest',\n",
       "   'value': 'RYY2yzWCjihuu2hmqFWu8qg7aUB3aPwR'}],\n",
       " 'moleculeType': {'coding': [{'system': 'http://hl7.org/fhir/sequence-type',\n",
       "    'code': 'dna',\n",
       "    'display': 'DNA Sequence'}]},\n",
       " 'location': [{'id': 'ga4gh:SL.3sVPpjle__QNFe8yGfLEja0R42VIZIp8',\n",
       "   'extension': [{'url': 'https://w3id.org/ga4gh/schema/vrs/2.0.1/json/SequenceLocation#properties/digest',\n",
       "     'valueString': '3sVPpjle__QNFe8yGfLEja0R42VIZIp8'}],\n",
       "   'sequenceLocation': {'sequenceContext': {'reference': '#vrs-location-sequenceReference',\n",
       "     'type': 'Sequence',\n",
       "     'display': 'VRS location.sequenceReference as contained FHIR Sequence'},\n",
       "    'coordinateInterval': {'coordinateSystem': {'system': {'coding': [{'system': 'http://loinc.org',\n",
       "         'code': 'LA30100-4',\n",
       "         'display': '0-based interval counting'}]}},\n",
       "     'startQuantity': {'value': 121530894.0},\n",
       "     'endQuantity': {'value': 121530895.0}}}}],\n",
       " 'representation': [{'focus': {'coding': [{'system': 'http://hl7.org/fhir/moleculardefinition-focus',\n",
       "      'code': 'allele-state'}]},\n",
       "   'code': [{'coding': [{'code': 'NC_000002.12:121530894:G:C',\n",
       "       'display': 'spdi'}]},\n",
       "    {'coding': [{'code': 'NC_000002.12:g.121530895G>C', 'display': 'hgvs.g'}]},\n",
       "    {'coding': [{'code': '2-121530895-G-C', 'display': 'gnomad'}]}],\n",
       "   'literal': {'value': 'C'}}]}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Display one example FHIR Allele Profile to show its structure.\n",
    "fhir_examples[0].model_dump()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
